compiler/AST/foralls.cpp

```diff
@@ -0,0 +1,236 @@
+    UPDATE(leadIdx,     newSE);
+    UPDATE(leadIdxCopy, newSE);
+  }
+  
+  // nothing matched
+#undef UPDATE
+  return false;
+}
+
+// This is intended to mimick Expr::remove(), without 'this' being an Expr.
+void ForallIntents::removeFI(Expr* parentB) {
+  // If this fails need to use trace_remove() instead of remove_help()
+  // - see Expr::remove().
+  INT_ASSERT(parentB->parentSymbol);
+
+  // "Remove" all ASTs that 'this' contains.
+#define REMOVE(dest) if (dest) remove_help(dest, 'r')
+
+  for_vector(Expr, var, fiVars) REMOVE(var);
+  for_riSpecs_vector(ri, this)  REMOVE(ri);
+  REMOVE(iterRec);
+  REMOVE(leadIdx);
+  REMOVE(leadIdxCopy);
+
+#undef REMOVE    
+}
+
+void ForallIntents::verifyFI(BlockStmt* parentB) {
+  Expr* parentE = (Expr*)parentB;
+  int nv = numVars();
+  INT_ASSERT((int)(fiVars.size())   == nv);
+  INT_ASSERT((int)(fIntents.size()) == nv);
+  INT_ASSERT((int)(riSpecs.size())  == nv);
+
+  for (int i = 0; i < nv; i++) {
+    Expr* fiVar = fiVars[i];
+    if (SymExpr* fiVarSE = toSymExpr(fiVar)) {
+      INT_ASSERT(isVarSymbol(fiVarSE->symbol()) ||
+                 isArgSymbol(fiVarSE->symbol()));  // no modules, fns, etc.
+    } else {
+      // fiVars[i] is either resolved or unresolved sym expr; never NULL.
+      INT_ASSERT(isUnresolvedSymExpr(fiVar));
+      // These should be resolved during scopeResolve.
+      INT_ASSERT(!normalized);
+    }
+    verifyNotOnList(fiVar);
+    INT_ASSERT(fiVar->parentExpr == parentE);
+
+    Expr* ri = riSpecs[i];
+    INT_ASSERT(isReduce(i) == !!ri);
+    if (ri) {
+      // ri can be UnresolvedSymExpr, SymExpr, CallExpr, ... (?)
+      verifyNotOnList(ri);
+      INT_ASSERT(ri->parentExpr == parentE);
+    }
+  }
+
+  INT_ASSERT(!iterRec || iterRec->parentExpr == parentE);
+  INT_ASSERT(!leadIdx || leadIdx->parentExpr == parentE);
+  INT_ASSERT(!leadIdxCopy || leadIdxCopy->parentExpr == parentE);
+  verifyNotOnList(iterRec);
+  verifyNotOnList(leadIdx);
+  verifyNotOnList(leadIdxCopy);
+
+  // ForallIntents are gone during resolve().
+  INT_ASSERT(!resolved);
+}
+
+void ForallIntents::acceptFI(AstVisitor* visitor) {
+  visitor->visitForallIntents(this);
+}
+
+
+//
+// parser support
+// These functions report a user error for an unexpected intent.
+//
+
+static TFITag it2tfi(Expr* ref, IntentTag intent) {
+  switch (intent) {
+  case INTENT_IN:        return TFI_IN;
+  case INTENT_CONST:     return TFI_CONST;
+  case INTENT_CONST_IN:  return TFI_CONST_IN;
+  case INTENT_REF:       return TFI_REF;
+  case INTENT_CONST_REF: return TFI_CONST_REF;
+  case INTENT_BLANK:     return TFI_DEFAULT;
+
+  case INTENT_OUT:
+  case INTENT_INOUT:
+  case INTENT_PARAM:
+  case INTENT_TYPE:
+    USR_FATAL_CONT(ref, "%s is not supported in a 'with' clause",
+                        intentDescrString(intent));
+    return TFI_DEFAULT;
+  }
+  INT_ASSERT(false); // unexpected IntentTag; 'intent' contains garbage?
+  return TFI_DEFAULT; // dummy
+}
+
+ForallIntents* new_ForallIntents() {
```

Comment:
Let's remove this function and just call the constructor directly.  Currently some places call this and some others call ForallIntents() directly.